ipa/verifier.go

```diff
@@ -48,34 +48,27 @@ func CheckIPAProof(transcript *common.Transcript, ic *IPAConfig, commitment band
 		commitment = commit([]banderwagon.Element{commitment, L, R}, []fr.Element{fr.One(), x, xInv})
 	}
 
-	current_basis := ic.SRSPrecompPoints.SRS
-
-	for i := 0; i < len(challenges); i++ {
-
-		G_L, G_R := splitPoints(current_basis)
-
-		b_L, b_R := splitScalars(b)
-
-		xInv := challenges_inv[i]
-
-		b = foldScalars(b_L, b_R, xInv)
-		current_basis = foldPoints(G_L, G_R, xInv)
+	g := ic.SRSPrecompPoints.SRS
+
+	// We compute the folding-scalars for g and b.
+	foldingScalars := make([]fr.Element, len(g))
+	for i := 0; i < len(g); i++ {
+		scalar := fr.One()
+
+		for challengeIdx := 0; challengeIdx < len(challenges); challengeIdx++ {
+			if i&(1<<(7-challengeIdx)) > 0 {
+				scalar.Mul(&scalar, &challenges_inv[challengeIdx])
+			}
+		}
+		foldingScalars[i] = scalar
 	}
-
-	if len(b) != len(current_basis) {
-		panic("reduction was not correct")
-	}
-
-	if len(b) != 1 {
-		panic("`b` and `G` should be 1")
-	}
-
-	b0 := b[0]
+	g0 := multiScalar(g, foldingScalars)
+	b0 := InnerProd(b, foldingScalars)
```

Comment:
After we have each final scalar for each vector element, we calculate the "final" `g0` and `b0` with a MSM and inner product respectively.